### DISCUSSION TITLE
Implement a "on-demand" mode on harbor streaming

### DISCUSSION THREAD

**Discussion by scd75** [savonet/liquidsoap#1978] [Ideas]
**Is your feature request related to a problem? Please describe.**
I use Liquidsoap to transcode on-the-fly internet radios whose codec format is not supported by my listening device. As an example, I transcode Radio-France streams (eg. http://icecast.radiofrance.fr/francemusique-hifi.aac) to an OGG stream, which I can then listen with my hi-fi network receiver. Yet I did not find any good solution for this transcoding to run only when my hi-fi device is actually listening. It means liquidsoap constantly listens and download the external stream, regardless of the end devices listening status. This implies unnecessary data and power consumption. This problem becomes even bigger when performing this operation for multiple streams (while actually listening to only 1 at the same time with my hifi device).

**Describe the solution you'd like**
I would like liquidsoap to only perform the decoding operation (and download of the initial stream) when my hi-fi device actually listens to the transcoded stream. I guess this implies adding a "on-demand" feature to the harbor streaming framework. Would also be great to have it on the icecast streaming framework, but not sure if it is possible as it would imply having to get feedback from icecast server.

**Describe alternatives you've considered**
as discussed with @toots , tried to have a URL redirection when connection to harbor with my hifi device, but did not work well.

**Additional context**
I guess Icecast's "on-demand" implementation for relayed stream would be a good starting point to consider

**Comment by toots** [comment]
Hi! 

Thanks for this detailed report. There are a couple of things worth noting here:
- We now have support for inline decoding using `ffmpeg` in `master`. This means that we can received an encoded stream and only decode it when the source using it is active!
- How do you receive the data? Harbor is usually receiving the stream so it must be sent from another place right?

**Comment by scd75** [comment]
Hello,

I am not sure I get how to use ffmpeg in my context? Plus, would it work with an AAC-encoded stream?
As for how i receive the data, I use input.http as a source, then I output it through harbor. Then my hi-fi receiver eventually connects to the mounted uri...
here is my setup:
```
source = input.http('http://icecast.radiofrance.fr/francemusique-hifi.aac')
output.harbor(%vorbis(samplerate=48000, channels=2, quality=0.6),
              fallible=true,
              mount="test-uri.ogg",
              port=8000,
              source)
````

**Comment by toots** [comment]
Ffmpeg would work with any format accepted by the underlying `libavcodec`/`libavformat` library. It is still an experimental feature, it was just added this week.

I understand your feature request now! There is still a little more that needs to be implemented. I'll see if I can get to it as part of liquidsoap 2. Thanks for suggesting!

**Comment by scd75** [comment]
Hi @toots , could you make any progress on this?
Thanks!

**Comment by toots** [comment]
Ok, I think that you should be able to do this as follows:
```
s = input.http("http://icecast.radiofrance.fr/francemusique-hifi.aac")

s = ffmpeg.encode.audio(
  %ffmpeg(
    %audio(codec="libvorbis", ...)
  ), s)

default = single("/path/to/vorbis.ogg")

should_decode = ref(false)

s = switch(track_sensitive=false, [
  ({!should_decode}, s),
  ({true},default)
])

def on_connect(_) =
  should_decode := true
end

def on_disconnect() =
  should_decode := false
end

output.harbor(%ffmpeg(format="ogg", %audio.copy), on_connect=on_connect, on_disconnect=on_disconnect, ...)
```
Does that make sense? The `input.http` source is decoded/encoded inline so, if it is not used in the `switch`, it should not generate CPU usage. Meanwhile, the `vorbis.ogg` file is decoded as encoded data so it should not generate CPU usage either.

**Reply by scd75** [reply]
hi @toots , aside of CPU usage, I would want to make sure that no data is downloaded when no listening device is connected. I that the case in your proposed solution?
Thanks!

**Reply by toots** [reply]
The source returned by `input.http` has a `start` and `stop` method that should be able to use for that:
```ruby
http_input = input.http("http://icecast.radiofrance.fr/francemusique-hifi.aac")

s = ffmpeg.encode.audio(
  %ffmpeg(
    %audio(codec="libvorbis", ...)
  ), http_input)

default = single("/path/to/vorbis.ogg")

should_decode = ref(false)

s = switch(track_sensitive=false, [
  ({!should_decode}, s),
  ({true},default)
])

def on_connect(_) =
  should_decode := true
  http_input.start()
end

def on_disconnect() =
  should_decode := false
  http_input.stop()
end

output.harbor(%ffmpeg(format="ogg", %audio.copy), on_connect=on_connect, on_disconnect=on_disconnect, ...)
```
